### ISSUE TITLE
How do/can atomic upserts work?

### ISSUE THREAD

**Issue by chrisfosterelli** [rethinkdb/rethinkdb#4581]
Hi guys!

So, to quote the [RethinkDB documentation](http://rethinkdb.com/docs/architecture/#query-execution):

> Currently, values obtained by executing JavaScript code, random values, and values obtained as a result of a subquery (e.g. incrementing the value of an attribute by the value of an attribute in a different document) cannot be performed atomically. If the user runs a query that cannot be executed atomically, by default RethinkDB will throw an error

When I attempt to `update`, I get an error _as expected_. However, this feature behaves counter-intuitively when I begin doing upserts:

```javascript
r.table('FruitCollection').insert({ 
  id : 1, 
  fruits : r.table('otherCollection').get(1)('fruits').count()
}, { 
  conflict : 'update' 
});
```

As I understand it, this operation **is not atomic when updating** based on the documentation. Although, it doesn't throw an error which suggests it is atomic. However, what about this operation:

```javascript
r.table('FruitCollection').insert({ 
  id : 1, 
  fruits : r.table('FruitCollection').get(1)('fruits').default([]).append('orange')
}, { 
  conflict : 'update' 
});
```

This is far more confusing as well. It uses a subquery, which should not be atomic, however it only touches one document, which is supposed to be atomic. It doesn't throw an error, so I would be led to believe this is atomic, but neither does the first query so the behaviour is very confusing.

What I would really like to do is this:

```javascript
r.table('FruitCollection').insert({ 
  id : 1, 
  fruits : r.row.default([]).append('orange')
}, { 
  conflict : 'update' 
});
```

But `r.row` is not supported when doing `insert`.

1. Should these first two queries throw a non-atomic error and require a `nonAtomic` flag?
2. Does it make sense to add `r.row` support to `insert`?
3. If not, is there no way to atomically upsert using append/add/etc in Rethinkdb?

I propose adding more clear documentation on atomicity to the pages on `insert`, and perhaps changing the behaviour to throw an error if it is not atomic. Thanks for all your hard work on Rethink!

**Comment by let4be** [comment]
There is also another attempt to make atomic upsert, but it looks suspicious in terms of atomicity:
http://stackoverflow.com/questions/24306933/how-to-make-a-rethinkdb-atomic-update-if-document-exists-insert-otherwise

**Comment by let4be** [comment]
From my experience 1 and 2 behaves atomically, results are as expected not like when doing exists check and then conditional update/insert locally
But would be nice to hear from devs regarding this

**Comment by chrisvariety** [comment]
This document may help a ilttle, from the 2.1 beta: http://docs.rethinkdb.com/2.1/docs/consistency/

Check "Linearizability and atomicity guarantees"

**Comment by danielmewes** [comment]
@chrisfosterelli Thanks for bringing this to our attention. The FAQ is indeed pretty misleading when it talks of "If the user runs a query that cannot be executed atomically, by default RethinkDB will throw an error". I'll open a separate issue in our docs repository https://github.com/rethinkdb/docs/issues/ so we can improve it.
What the documentation describes only applies to `update` and `replace`. `insert` was never intended to perform arbitrary mutations, so it doesn't take a function either (neither atomic nor non-atomic).

The first two queries are *not* atomic. As soon as you involve a subquery in `insert` it will not be atomic anymore.

The only way to do this operation atomically at the moment is through `replace`:
```
r.table('FruitCollection').get(1).replace({ 
  id : 1, 
  fruits : r.row.default([]).append('orange')
});
```

`replace` will insert a new document if the key doesn't exist.

**Comment by danielmewes** [comment]
Make the FAQ not misleading: https://github.com/rethinkdb/docs/issues/834

**Comment by danielmewes** [comment]
Also just opened https://github.com/rethinkdb/rethinkdb/issues/4582 to improve the atomicity check for `update` and `replace` (this wouldn't have helped in this specific case with `insert`, but your example uncovered a related problem with those operations).

**Comment by chrisfosterelli** [comment]
Thanks @danielmewes! I think updating the documentation to specify that its only referring to `update`/`replace` operations when it throws an error would make this _way more straightforward_ and would've cleared up my confusion. It would be even better if it could be worded in a way that made it clear subqueries like I was trying to do with `insert` would _not_ be atomic.

In regards to #4582, that would a really neat feature as well and would make the behaviour much more intuitive.

Also, thanks for providing the actual code I was looking for that does an atomic upsert with `append`. That's not the intuitive approach I would've guessed, but it's great to know it's possible :)

It seems like my question is resolved, and the rethinkdb/docs#834 and #4582 have been created to deal with the individual issues. OK to close this issue then?

**Comment by danielmewes** [comment]
@chrisfosterelli Thanks for your feedback. 
Closing this as suggested. Going to track the improvements in the separate issues.
